dynawo/sources/Models/Modelica/Dynawo/NonElectrical/Blocks/Continuous/IntegratorVariableLimits.mo

```diff
@@ -0,0 +1,110 @@
+within Dynawo.NonElectrical.Blocks.Continuous;
+
+/*
+* Copyright (c) 2022, RTE (http://www.rte-france.com)
+* See AUTHORS.txt
+* All rights reserved.
+* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, you can obtain one at http://mozilla.org/MPL/2.0/.
+* SPDX-License-Identifier: MPL-2.0
+*
+* This file is part of Dynawo, an hybrid C++/Modelica open source suite
+* of simulation tools for power systems.
+*/
+
+model IntegratorVariableLimits "Integrator with limited value of output (variable limits) and freeze"
+  import Modelica;
+  import Dynawo.Types;
+
+  extends Modelica.Blocks.Interfaces.SISO(y(start = Y0));
+
+  parameter Boolean DefaultLimitMax = true "If limitMin > limitMax : if true, y = limitMax, if false, y = limitMin";
+  parameter Types.PerUnit K = 1 "Integrator gain";
+  parameter Types.PerUnit Kf = 0 "Feedback gain";
+  parameter Types.PerUnit Tol = 0 "Tolerance on limit crossing";
+
+  Modelica.Blocks.Interfaces.RealInput limitMax "Connector of Real input signal used as maximum of input u" annotation(
+    Placement(transformation(extent={{-140,60},{-100,100}})));
+  Modelica.Blocks.Interfaces.RealInput limitMin "Connector of Real input signal used as minimum of input u" annotation(
+    Placement(transformation(extent={{-140,-100},{-100,-60}})));
+
+  parameter Boolean Frozen0 = false "If true, integration is frozen at initial time";
+  parameter Types.PerUnit Y0 = 0 "Initial or guess value of output (must be in the limits limitMin .. limitMax)";
+
+protected
+  Boolean isFrozen(start = Frozen0) "If true, integration is frozen";
+  Boolean keepFreezing(start = Frozen0) "If true, integration stays frozen";
+  Boolean startFreezing(start = Frozen0) "If true, integration becomes frozen";
+  Types.PerUnit v "Integrator input accounting for the feedback signal but not for the freeze";
+  Types.PerUnit w(start = Y0) "Non-limited integrator output";
+
+equation
+  v = K * u + Kf * (y - w);
+
+  startFreezing = (w < limitMin and v < 0) or (w > limitMax and v > 0);
+  keepFreezing = ((w < limitMin + Tol and v < 0) or (w > limitMax - Tol and v > 0)) and pre(isFrozen);
```

Comment:
same here : the integrator unfreezes only when it is far enough from the limits

Reply:
Integration is unfrozen if the integrator output is within the limits and far enough from them or if the integrator input tends to bring the output back within the limits

Reply:
ok